src/jit/morph.cpp

```diff
@@ -12829,6 +12829,26 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
             break;
 #endif
 
+        case GT_ARR_LENGTH:
+            // Morph "constant_string".Length to CNS_INT(15)
+            if (op1->OperIs(GT_IND) && fgGlobalMorph)
+            {
+                GenTree* strCon = op1->AsOp()->gtGetOp1();
+                if (strCon->OperIs(GT_CNS_INT) && strCon->IsIconHandle(GTF_ICON_STR_HDL))
+                {
+                    CORINFO_String* asString = *reinterpret_cast<CORINFO_String**>(strCon->AsIntCon()->IconValue());
```

Comment:
Hmm, I remember looking into this in the past and thinking that the JIT-VM interface would need to be extended to obtain this information. But I don't remember why, perhaps it wasn't clear if it's appropriate to use `CORINFO_String` like this.

Reply:
It is not ok to use `CORINFO_String` like this. The string is not pinned and the GC can move it while the JIT is looking at the Length field. It would result into GC hole / intermittent silent bad codegen.

Reply:
Oh I suspected that, I hope I won't have to modify vm API for it, looking at the importation now to find out if I can save string length to GenTreeStrCon (new field)

Reply:
I do not think you would be able to do this without modifying JIT/EE interface

Reply:
@jkotas ah I see, I guess it should be something similar to `isValidStringRef`

Reply:
@jkotas I tried to copy-paste `isValidStringRef` method and introduced `getStringLength` should work now I hope...

Reply:
If a new function is added to the JIT-VM interface would it make sense to also return the string characters somehow? Perhaps in the future we could add some other optimizations around constant strings such as generating special code for `a.Equals("ABCD")`.

Reply:
>  I remember looking into this in the past and thinking that the JIT-VM interface would need to be extended to obtain this information. But I don't remember why, 

"Jit doesn't recognise ldstr as non-null"? https://github.com/dotnet/coreclr/issues/22511#issuecomment-462440743 

> That seems to be unexpectedly complicated, the VM keeps the JIT in the dark about the string length. The JIT interface will probably need to be extended.